packages/dd-trace/src/debugger/devtools_client/status.js

```diff
@@ -0,0 +1,82 @@
+'use strict'
+
+const config = require('./config')
+const request = require('../../exporters/common/request')
+const FormData = require('../../exporters/common/form-data')
+const log = require('../../log')
+
+module.exports = {
+  ackReceived,
+  ackInstalled,
+  ackEmitting,
+  ackError
+}
+
+const ddsource = 'dd_debugger'
+const service = config.service
+const runtimeId = config.runtimeId
+
+const STATUSES = {
+  RECEIVED: 'RECEIVED',
+  INSTALLED: 'INSTALLED',
+  EMITTING: 'EMITTING',
+  ERROR: 'ERROR',
+  BLOCKED: 'BLOCKED' // TODO: Implement once support for allow list, deny list or max probe limit has been added
+}
+
+function ackReceived ({ id: probeId, version }) {
```

Comment:
I think there are two features missing from probe status implementation:
1. don't repeat status messages. for example, only send emitting/error once.
2. once every hour, send probe status for all probes on the list. this should basiclly emit the last status of that probe.

If you ask why we are doing 2 - well, that because the UI would like to query the events from the last hour and get all the probe status - even if they where sent 3 days ago.

Reply:
> 1. don't repeat status messages. for example, only send emitting/error once.

So if a probe is first created and then later modified, we should only send the `RECEIVED` payload once? How will the UI then know that the RC message to modify the probe was received?

So for a givne unique probe id, we never want to send the same diagnostic message more than once? E.g. if a probe is first created and then later modified, we should only send the `RECEIVED` payload once? And if the same probe is being triggered multiple times, we should also only send the `EMITTING` payload the first time it was triggered?

> 2. once every hour, send probe status for all probes on the list. this should basiclly emit the last status of that probe.
>
> If you ask why we are doing 2 - well, that because the UI would like to query the events from the last hour and get all the probe status - even if they where sent 3 days ago.

I think this state management should really be handled by the backend database and not delegated to the tracers - But I have a feeling you agree with me already 😅 In any case, I'll implement this and then try to forget about it as soon as possible 😭

Reply:
Follow up question: If a given probe experiences more than one unique error, we still only want to send the first one we see within a given hour? So if we in hour one, see 3 errors, all unique, we only send the first one, then one hour and one second after we saw the first error, a new error happens, which we sent, effectively dropping the two inbetween?

Reply:
Yeah, you are right - we send a single probe per probe-id+version combo. so when a probe get updated, the probe-version increase by 1 and then you emit new RECEIVED message. then that status can be upgraded to INSTALLED and EMITTING (or ERROR/BLOCKED on the sad path).

Yup. the backend can store that at the backend - but that train left station long time ago

Reply:
I've now implemented a check to ensure we don't send the same acknowledge-request twice within the same hour. I have specifically _not_ implemented a timer to resubmit the status of all probes ones an hour as this is no longer strictly needed according to @cimi. I've added this topic to the agent sync agenda backlog.